File: MaxStack.js

/*
 The problem statement guaranatees that there will be at least one element 
 in the stack when pop, top, peekMax, or popMax is called.
 */
var MaxStack = function () {

    this.indexMaxElement = 0;
    this.container = [];
    this.indexTop = -1;
    ;
};

/** 
 * @param {number} x
 * @return {void}
 */
MaxStack.prototype.push = function (x) {
    this.container[++this.indexTop] = x;
    if (this.indexTop > 0 && this.container[this.indexMaxElement] <= x) {
        this.indexMaxElement = this.indexTop;
    }
};

/**
 * @return {number}
 */
MaxStack.prototype.pop = function () {
    if (this.indexMaxElement === this.indexTop) {
        this.update_indexMaxElement();
    }
    return this.container[this.indexTop--];
};

/**
 * @return {number}
 */
MaxStack.prototype.top = function () {
    return this.container[this.indexTop];
};

/**
 * @return {number}
 */
MaxStack.prototype.peekMax = function () {
    return this.container[this.indexMaxElement];
};

/**
 * @return {number}
 */
MaxStack.prototype.popMax = function () {

    let maxElement = this.container[this.indexMaxElement];

    //Move all elements after 'indexMaxElement' one step to the left.
    while (this.indexMaxElement < this.indexTop) {
        this.container[this.indexMaxElement] = this.container[++this.indexMaxElement];
    }
    this.update_indexMaxElement();
    this.indexTop--;
    return maxElement;
};


MaxStack.prototype.update_indexMaxElement = function () {
    this.indexMaxElement = 0;
    let indexNextMaxElement = 0;
    while (indexNextMaxElement < this.indexTop) {
        if (this.container[indexNextMaxElement] >= this.container[this.indexMaxElement]) {
            this.indexMaxElement = indexNextMaxElement;
        }
        indexNextMaxElement++;
    }
};

File: MaxStack.cpp

#include<array>
using namespace std;


const int MAXIMUM_CALLS_TO_STACK = 10000;
int container[MAXIMUM_CALLS_TO_STACK] = { 0 };

class MaxStack {
public:

	int indexMaxElement;
	int indexTop;

	/*
	The problem statement guaranatees that there will be at least one element
	in the stack when pop, top, peekMax, or popMax is called.
	*/
	MaxStack() {
		indexMaxElement = 0;
		indexTop = -1;
	}

	void push(int x) {
		container[++indexTop] = x;
		if (indexTop > 0 && container[indexMaxElement] <= x) {
			indexMaxElement = indexTop;
		}
	}

	int pop() {
		if (indexMaxElement == indexTop) {
			update_indexMaxElement();
		}
		return container[indexTop--];
	}

	int top() {
		return container[indexTop];
	}

	int peekMax() {
		return container[indexMaxElement];
	}

	int popMax() {

		int maxElement = container[indexMaxElement];

		//Move all elements after 'indexMaxElement' one step to the left.
		while (indexMaxElement < indexTop) {
			container[indexMaxElement] = container[indexMaxElement + 1];
			++indexMaxElement;
		}

		update_indexMaxElement();
		indexTop--;
		return maxElement;
	}

	void update_indexMaxElement() {
		indexMaxElement = 0;
		int indexNextMaxElement = 0;
		while (indexNextMaxElement < indexTop) {
			if (container[indexNextMaxElement] >= container[indexMaxElement]) {
				indexMaxElement = indexNextMaxElement;
			}
			indexNextMaxElement++;
		}
	}
};
